<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Calculate alignments with a generalized variant of the Smith Waterman algorithm. Complexity: O(n * m) for strings of length m and n."><meta name="keywords" content="rust, rustlang, rust-lang, pairwise"><title>bio::alignment::pairwise - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../../../normalize.css"><link rel="stylesheet" type="text/css" href="../../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../../../dark.css" disabled><link rel="stylesheet" type="text/css" href="../../../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../../../storage.js"></script><script src="../../../crates.js"></script><script defer src="../../../main.js"></script>
    <noscript><link rel="stylesheet" href="../../../noscript.css"></noscript><link rel="icon" href="https://raw.githubusercontent.com/rust-bio/rust-bio/master/img/bioferris.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../../bio/index.html"><div class="logo-container"><img src="https://raw.githubusercontent.com/rust-bio/rust-bio/master/img/bioferris.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../../../bio/index.html"><div class="logo-container">
                    <img src="https://raw.githubusercontent.com/rust-bio/rust-bio/master/img/bioferris.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Module pairwise</a></h2><div class="sidebar-elems"><section><div class="block"><ul><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#constants">Constants</a></li><li><a href="#traits">Traits</a></li></ul></div></section><div id="sidebar-vars" data-name="pairwise" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../../../bio/index.html">
                        <img src="https://raw.githubusercontent.com/rust-bio/rust-bio/master/img/bioferris.svg" alt="logo"></a><nav class="sub"><div class="theme-picker hidden"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="22" height="22" alt="Pick another theme!" src="../../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../../wheel.svg"></a></div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Module <a href="../../index.html">bio</a>::<wbr><a href="../index.html">alignment</a>::<wbr><a class="mod" href="#">pairwise</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../../../src/bio/alignment/pairwise/mod.rs.html#6-1716">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Calculate alignments with a generalized variant of the Smith Waterman algorithm.
Complexity: O(n * m) for strings of length m and n.</p>
<p>For quick computation of alignments and alignment scores there are 6 simple functions.</p>
<h2 id="example"><a href="#example">Example</a></h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use</span> <span class="ident">bio::alignment::pairwise</span>::<span class="kw-2">*</span>;
<span class="kw">use</span> <span class="ident">bio::alignment::AlignmentOperation</span>::<span class="kw-2">*</span>;
<span class="kw">use</span> <span class="ident">bio::scores::blosum62</span>;

<span class="kw">let</span> <span class="ident">x</span> <span class="op">=</span> <span class="string">b&quot;ACCGTGGAT&quot;</span>;
<span class="kw">let</span> <span class="ident">y</span> <span class="op">=</span> <span class="string">b&quot;AAAAACCGTTGAT&quot;</span>;
<span class="kw">let</span> <span class="ident">score</span> <span class="op">=</span> <span class="op">|</span><span class="ident">a</span>: <span class="ident">u8</span>, <span class="ident">b</span>: <span class="ident">u8</span><span class="op">|</span> <span class="kw">if</span> <span class="ident">a</span> <span class="op">==</span> <span class="ident">b</span> { <span class="number">1i32</span> } <span class="kw">else</span> { <span class="op">-</span><span class="number">1i32</span> };
<span class="comment">// gap open score: -5, gap extension score: -1</span>
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">aligner</span> <span class="op">=</span> <span class="ident">Aligner::with_capacity</span>(<span class="ident">x</span>.<span class="ident">len</span>(), <span class="ident">y</span>.<span class="ident">len</span>(), <span class="op">-</span><span class="number">5</span>, <span class="op">-</span><span class="number">1</span>, <span class="kw-2">&amp;</span><span class="ident">score</span>);
<span class="kw">let</span> <span class="ident">alignment</span> <span class="op">=</span> <span class="ident">aligner</span>.<span class="ident">semiglobal</span>(<span class="ident">x</span>, <span class="ident">y</span>);
<span class="comment">// x is global (target sequence) and y is local (reference sequence)</span>
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">ystart</span>, <span class="number">4</span>);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">xstart</span>, <span class="number">0</span>);
<span class="macro">assert_eq!</span>(
    <span class="ident">alignment</span>.<span class="ident">operations</span>,
    [<span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Subst</span>, <span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Match</span>]
);

<span class="comment">// You can use predefined scoring matrices such as BLOSUM62</span>
<span class="kw">let</span> <span class="ident">x</span> <span class="op">=</span> <span class="string">b&quot;LSPADKTNVKAA&quot;</span>;
<span class="kw">let</span> <span class="ident">y</span> <span class="op">=</span> <span class="string">b&quot;PEEKSAV&quot;</span>;
<span class="comment">// gap open score: -10, gap extension score: -1</span>
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">aligner</span> <span class="op">=</span> <span class="ident">Aligner::with_capacity</span>(<span class="ident">x</span>.<span class="ident">len</span>(), <span class="ident">y</span>.<span class="ident">len</span>(), <span class="op">-</span><span class="number">10</span>, <span class="op">-</span><span class="number">1</span>, <span class="kw-2">&amp;</span><span class="ident">blosum62</span>);
<span class="kw">let</span> <span class="ident">alignment</span> <span class="op">=</span> <span class="ident">aligner</span>.<span class="ident">local</span>(<span class="ident">x</span>, <span class="ident">y</span>);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">xstart</span>, <span class="number">2</span>);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">xend</span>, <span class="number">9</span>);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">ystart</span>, <span class="number">0</span>);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">yend</span>, <span class="number">7</span>);
<span class="macro">assert_eq!</span>(
    <span class="ident">alignment</span>.<span class="ident">operations</span>,
    [<span class="ident">Match</span>, <span class="ident">Subst</span>, <span class="ident">Subst</span>, <span class="ident">Match</span>, <span class="ident">Subst</span>, <span class="ident">Subst</span>, <span class="ident">Match</span>]
);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">score</span>, <span class="number">16</span>);

<span class="comment">// If you don&#39;t know sizes of future sequences, you could</span>
<span class="comment">// use Aligner::new().</span>
<span class="comment">// Global alignment:</span>
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">aligner</span> <span class="op">=</span> <span class="ident">Aligner::new</span>(<span class="op">-</span><span class="number">5</span>, <span class="op">-</span><span class="number">1</span>, <span class="kw-2">&amp;</span><span class="ident">score</span>);
<span class="kw">let</span> <span class="ident">x</span> <span class="op">=</span> <span class="string">b&quot;ACCGTGGAT&quot;</span>;
<span class="kw">let</span> <span class="ident">y</span> <span class="op">=</span> <span class="string">b&quot;AAAAACCGTTGAT&quot;</span>;
<span class="kw">let</span> <span class="ident">alignment</span> <span class="op">=</span> <span class="ident">aligner</span>.<span class="ident">global</span>(<span class="ident">x</span>, <span class="ident">y</span>);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">ystart</span>, <span class="number">0</span>);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">xstart</span>, <span class="number">0</span>);
<span class="macro">assert_eq!</span>(<span class="ident">aligner</span>.<span class="ident">local</span>(<span class="ident">x</span>, <span class="ident">y</span>).<span class="ident">score</span>, <span class="number">7</span>);

<span class="comment">// In addition to the standard modes (Global, Semiglobal and Local), a custom alignment</span>
<span class="comment">// mode is supported which supports a user-specified clipping penalty. Clipping is a</span>
<span class="comment">// special boundary condition where you are allowed to clip off the beginning/end of</span>
<span class="comment">// the sequence for a fixed penalty. As a starting example, we can use the custom mode</span>
<span class="comment">// for achieving the three standard modes as follows.</span>

<span class="comment">// scoring for semiglobal mode</span>
<span class="kw">let</span> <span class="ident">scoring</span> <span class="op">=</span> <span class="ident">Scoring::new</span>(<span class="op">-</span><span class="number">5</span>, <span class="op">-</span><span class="number">1</span>, <span class="kw-2">&amp;</span><span class="ident">score</span>) <span class="comment">// Gap open, gap extend and match score function</span>
    .<span class="ident">xclip</span>(<span class="ident">MIN_SCORE</span>) <span class="comment">// Clipping penalty for x set to &#39;negative infinity&#39;, hence global in x</span>
    .<span class="ident">yclip</span>(<span class="number">0</span>); <span class="comment">// Clipping penalty for y set to 0, hence local in y</span>
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">aligner</span> <span class="op">=</span> <span class="ident">Aligner::with_scoring</span>(<span class="ident">scoring</span>);
<span class="kw">let</span> <span class="ident">alignment</span> <span class="op">=</span> <span class="ident">aligner</span>.<span class="ident">custom</span>(<span class="ident">x</span>, <span class="ident">y</span>); <span class="comment">// The custom aligner invocation</span>
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">ystart</span>, <span class="number">4</span>);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">xstart</span>, <span class="number">0</span>);
<span class="comment">// Note that in the custom mode, the clips are explicitly mentioned in the operations</span>
<span class="macro">assert_eq!</span>(
    <span class="ident">alignment</span>.<span class="ident">operations</span>,
    [
        <span class="ident">Yclip</span>(<span class="number">4</span>),
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Subst</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>
    ]
);

<span class="comment">// scoring for global mode</span>
<span class="comment">// scoring can also be created using from_scores if the match and mismatch scores are constants</span>
<span class="kw">let</span> <span class="ident">scoring</span> <span class="op">=</span> <span class="ident">Scoring::from_scores</span>(<span class="op">-</span><span class="number">5</span>, <span class="op">-</span><span class="number">1</span>, <span class="number">1</span>, <span class="op">-</span><span class="number">1</span>) <span class="comment">// Gap open, extend, match, mismatch score</span>
    .<span class="ident">xclip</span>(<span class="ident">MIN_SCORE</span>) <span class="comment">// Clipping penalty for x set to &#39;negative infinity&#39;, hence global in x</span>
    .<span class="ident">yclip</span>(<span class="ident">MIN_SCORE</span>); <span class="comment">// Clipping penalty for y set to &#39;negative infinity&#39;, hence global in y</span>
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">aligner</span> <span class="op">=</span> <span class="ident">Aligner::with_scoring</span>(<span class="ident">scoring</span>);
<span class="kw">let</span> <span class="ident">alignment</span> <span class="op">=</span> <span class="ident">aligner</span>.<span class="ident">custom</span>(<span class="ident">x</span>, <span class="ident">y</span>); <span class="comment">// The custom aligner invocation</span>
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">ystart</span>, <span class="number">0</span>);
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">xstart</span>, <span class="number">0</span>);
<span class="comment">// Note that in the custom mode, the clips are explicitly mentioned in the operations</span>
<span class="macro">assert_eq!</span>(
    <span class="ident">alignment</span>.<span class="ident">operations</span>,
    [<span class="ident">Del</span>, <span class="ident">Del</span>, <span class="ident">Del</span>, <span class="ident">Del</span>, <span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Subst</span>, <span class="ident">Match</span>, <span class="ident">Match</span>, <span class="ident">Match</span>]
);

<span class="comment">// Similarly if the clip penalties are both set to 0, we have local alignment mode. The scoring</span>
<span class="comment">// struct also lets users set different penalties for prefix/suffix clipping, thereby letting</span>
<span class="comment">// users have the flexibility to create a wide variety of boundary conditions. The xclip() and</span>
<span class="comment">// yclip() methods sets the prefix and suffix penalties to be equal. The scoring struct can be</span>
<span class="comment">// explicitly constructed for full flexibility.</span>

<span class="comment">// The following example considers a modification of the semiglobal mode where you are allowed</span>
<span class="comment">// to skip a prefix of the target sequence x, for a penalty of -10, but you have to consume</span>
<span class="comment">// the rest of the string in the alignment</span>

<span class="kw">let</span> <span class="ident">scoring</span> <span class="op">=</span> <span class="ident">Scoring</span> {
    <span class="ident">gap_open</span>: <span class="op">-</span><span class="number">5</span>,
    <span class="ident">gap_extend</span>: <span class="op">-</span><span class="number">1</span>,
    <span class="ident">match_fn</span>: <span class="op">|</span><span class="ident">a</span>: <span class="ident">u8</span>, <span class="ident">b</span>: <span class="ident">u8</span><span class="op">|</span> <span class="kw">if</span> <span class="ident">a</span> <span class="op">==</span> <span class="ident">b</span> { <span class="number">1i32</span> } <span class="kw">else</span> { <span class="op">-</span><span class="number">3i32</span> },
    <span class="ident">match_scores</span>: <span class="prelude-val">Some</span>((<span class="number">1</span>, <span class="op">-</span><span class="number">3</span>)),
    <span class="ident">xclip_prefix</span>: <span class="op">-</span><span class="number">10</span>,
    <span class="ident">xclip_suffix</span>: <span class="ident">MIN_SCORE</span>,
    <span class="ident">yclip_prefix</span>: <span class="number">0</span>,
    <span class="ident">yclip_suffix</span>: <span class="number">0</span>,
};
<span class="kw">let</span> <span class="ident">x</span> <span class="op">=</span> <span class="string">b&quot;GGGGGGACGTACGTACGT&quot;</span>;
<span class="kw">let</span> <span class="ident">y</span> <span class="op">=</span> <span class="string">b&quot;AAAAACGTACGTACGTAAAA&quot;</span>;
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">aligner</span> <span class="op">=</span> <span class="ident">Aligner::with_capacity_and_scoring</span>(<span class="ident">x</span>.<span class="ident">len</span>(), <span class="ident">y</span>.<span class="ident">len</span>(), <span class="ident">scoring</span>);
<span class="kw">let</span> <span class="ident">alignment</span> <span class="op">=</span> <span class="ident">aligner</span>.<span class="ident">custom</span>(<span class="ident">x</span>, <span class="ident">y</span>);
<span class="macro">println!</span>(<span class="string">&quot;{}&quot;</span>, <span class="ident">alignment</span>.<span class="ident">pretty</span>(<span class="ident">x</span>, <span class="ident">y</span>));
<span class="macro">assert_eq!</span>(<span class="ident">alignment</span>.<span class="ident">score</span>, <span class="number">2</span>);
<span class="macro">assert_eq!</span>(
    <span class="ident">alignment</span>.<span class="ident">operations</span>,
    [
        <span class="ident">Yclip</span>(<span class="number">4</span>),
        <span class="ident">Xclip</span>(<span class="number">6</span>),
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Match</span>,
        <span class="ident">Yclip</span>(<span class="number">4</span>)
    ]
);</code></pre></div>
</div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="banded/index.html" title="bio::alignment::pairwise::banded mod">banded</a></div><div class="item-right docblock-short"><p>Banded Smith-Waterman alignment for fast comparison of long strings.
Use sparse dynamic programming to find a ‘backbone’ alignment from exact
k-mer matches, then compute the SW alignment in a ‘band’ surrounding the
backbone, with a configurable width w. This method is not guaranteed
to recover the Smith-Waterman alignment, but will usually find the same
alignment if a) there is a reasonable density of exact k-mer matches
between the sequences, and b) the width parameter w is larger than the
excursion of the alignment path from diagonal between successive kmer
matches.  This technique is employed in long-read aligners (e.g. BLASR and BWA)
to drastically reduce runtime compared to Smith Waterman.
Complexity roughly O(min(m,n) * w)</p>
</div></div></div><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Aligner.html" title="bio::alignment::pairwise::Aligner struct">Aligner</a></div><div class="item-right docblock-short"><p>A generalized Smith-Waterman aligner.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.MatchParams.html" title="bio::alignment::pairwise::MatchParams struct">MatchParams</a></div><div class="item-right docblock-short"><p>A concrete data structure which implements trait MatchFunc with constant
match and mismatch scores</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.Scoring.html" title="bio::alignment::pairwise::Scoring struct">Scoring</a></div><div class="item-right docblock-short"><p>Details of scoring are encapsulated in this structure.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="struct" href="struct.TracebackCell.html" title="bio::alignment::pairwise::TracebackCell struct">TracebackCell</a></div><div class="item-right docblock-short"><p>Packed representation of one cell of a Smith-Waterman traceback matrix.
Stores the I, D and S traceback matrix values in two bytes.
Possible traceback moves include : start, insert, delete, match, substitute,
prefix clip and suffix clip for x &amp; y. So we need 4 bits each for matrices I, D, S
to keep track of these 9 moves.</p>
</div></div></div><h2 id="constants" class="small-section-header"><a href="#constants">Constants</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="constant" href="constant.MIN_SCORE.html" title="bio::alignment::pairwise::MIN_SCORE constant">MIN_SCORE</a></div><div class="item-right docblock-short"><p>Value to use as a ‘negative infinity’ score. Should be close to <code>i32::MIN</code>,
but avoid underflow when used with reasonable scoring parameters or even
adding two negative infinities. Use ~ <code>0.4 * i32::MIN</code></p>
</div></div></div><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.MatchFunc.html" title="bio::alignment::pairwise::MatchFunc trait">MatchFunc</a></div><div class="item-right docblock-short"><p>Trait required to instantiate a Scoring instance</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../../../" data-current-crate="bio" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.62.1 (e092d0b6b 2022-07-16)" ></div>
</body></html>