<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Rust-bio, a bioinformatics library for Rust."><meta name="keywords" content="rust, rustlang, rust-lang, bio"><title>bio - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script src="../crates.js"></script><script defer src="../main.js"></script>
    <noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" href="https://raw.githubusercontent.com/rust-bio/rust-bio/master/img/bioferris.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../bio/index.html"><div class="logo-container"><img src="https://raw.githubusercontent.com/rust-bio/rust-bio/master/img/bioferris.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../bio/index.html"><div class="logo-container">
                    <img src="https://raw.githubusercontent.com/rust-bio/rust-bio/master/img/bioferris.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Crate bio</a></h2><div class="sidebar-elems"><div class="block"><ul><li class="version">Version 0.41.0</li><li><a id="all-types" href="all.html">All Items</a></li></div></ul><section><div class="block"><ul><li><a href="#modules">Modules</a></li></ul></div></section><div id="sidebar-vars" data-name="bio" data-ty="mod" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../bio/index.html">
                        <img src="https://raw.githubusercontent.com/rust-bio/rust-bio/master/img/bioferris.svg" alt="logo"></a><nav class="sub"><div class="theme-picker hidden"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="22" height="22" alt="Pick another theme!" src="../brush.svg"></button><div id="theme-choices" role="menu"></div></div><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">bio</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/bio/lib.rs.html#6-263">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h2 id="rust-bio-a-bioinformatics-library-for-rust"><a href="#rust-bio-a-bioinformatics-library-for-rust">Rust-bio, a bioinformatics library for Rust.</a></h2>
<p>This library provides implementations of many algorithms and data structures
that are useful for bioinformatics.
All provided implementations are rigorously tested via continuous
integration.</p>
<p>For <strong>getting started</strong> with using <code>rust-bio</code>, see <a href="#getting-started">the <code>Getting started</code> section below</a>.
For navigating the documentation of the available modules, see <a href="#modules">the <code>Modules</code> section below</a>.
If you want to contribute to <code>rust-bio</code>, see <a href="https://github.com/rust-bio/rust-bio#contribute">the <code>Contribute</code> section in the repo</a>.</p>
<p>Currently, rust-bio provides</p>
<ul>
<li>most major pattern matching algorithms,</li>
<li>a convenient alphabet implementation,</li>
<li>pairwise alignment,</li>
<li>suffix arrays,</li>
<li>the <a href="">Burrows-Wheeler-transform (BWT)</a></li>
<li>the <a href="https://doi.org/10.1109/SFCS.2000.892127">Full-text index in Minute space index (FM-index)</a>,</li>
<li>FMD-Index for finding supermaximal exact matches,</li>
<li>a q-gram index,</li>
<li>utilities to work with <a href="https://en.wikipedia.org/wiki/Position_weight_matrix">PSSMs</a>,</li>
<li>an open reading frame (ORF) search algorithm,</li>
<li>a rank/select data structure,</li>
<li><a href="https://github.com/serde-rs/serde">serde</a> support for all data structures when built with <code>nightly</code> feature,</li>
<li>readers and writers for FASTQ, FASTA and BED,</li>
<li>helper functions for combinatorics and dealing with log probabilities,</li>
<li>an implementation of the Hidden Markov Model and related algorithms.</li>
</ul>
<p>For reading and writing SAM/BAM/CRAM, VCF/BCF files or tabix indexed files, have a look at <a href="https://docs.rs/rust-htslib">rust-htslib</a>.</p>
<h2 id="getting-started"><a href="#getting-started">Getting started</a></h2>
<p>We explain how to use Rust-Bio step-by-step.
Users who already have experience with Rust can skip right to <a href="https://docs.rs/bio/#step-3-use-rust-bio-in-your-project">Step 3: Use Rust-Bio in your project</a>.
Users who already know <code>rust-bio</code> might want to jump right into the <a href="https://docs.rs/bio/#modules">modules docs</a></p>
<h3 id="step-1-setting-up-rust"><a href="#step-1-setting-up-rust">Step 1: Setting up Rust</a></h3>
<p>Rust can be installed following the instruction for <a href="https://rustup.rs/">rustup</a>.</p>
<h3 id="step-2-setting-up-a-new-rust-project"><a href="#step-2-setting-up-a-new-rust-project">Step 2: Setting up a new Rust project</a></h3>
<p>Since Rust-Bio is a library, you need to setup your own new Rust project to use Rust-Bio.
With Rust, projects and their dependencies are managed with the builtin package manager <a href="https://doc.rust-lang.org/cargo/index.html">Cargo</a>.
To create a new Rust project, issue</p>
<div class="example-wrap"><pre class="language-bash"><code>cargo new hello_world --bin
cd hello_world</code></pre></div>
<p>in your terminal. The flag <code>--bin</code> tells Cargo to create an executable project instead of a library.
In <a href="https://doc.rust-lang.org/nightly/book/hello-cargo.html#a-new-project">this section</a> of the Rust docs, you find details about what Cargo just created for you.</p>
<p>Your new project can be compiled with</p>
<div class="example-wrap"><pre class="language-bash"><code>cargo build</code></pre></div>
<p>If dependencies in your project are out of date, update with</p>
<div class="example-wrap"><pre class="language-bash"><code>cargo update</code></pre></div>
<p>Execute the compiled code with</p>
<div class="example-wrap"><pre class="language-bash"><code>cargo run</code></pre></div>
<p>If you are new to Rust, we suggest to proceed with <a href="https://www.rust-lang.org/learn">learning Rust</a> via the Rust docs.</p>
<h3 id="step-3-use-rust-bio-in-your-project"><a href="#step-3-use-rust-bio-in-your-project">Step 3: Use Rust-Bio in your project</a></h3>
<p>To use Rust-Bio in your Rust project, add the following to your <code>Cargo.toml</code></p>
<div class="example-wrap"><pre class="language-toml"><code>[dependencies]
bio = &quot;*&quot;</code></pre></div>
<p>and import the crate from your source code:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">extern</span> <span class="kw">crate</span> <span class="ident">bio</span>;</code></pre></div>
<h3 id="example-fm-index-and-fastq"><a href="#example-fm-index-and-fastq">Example: FM-index and FASTQ</a></h3>
<p>An example of using <code>rust-bio</code>:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="comment">// Import some modules</span>
<span class="kw">use</span> <span class="ident">bio::alphabets</span>;
<span class="kw">use</span> <span class="ident">bio::data_structures::bwt</span>::{<span class="ident">bwt</span>, <span class="ident">less</span>, <span class="ident">Occ</span>};
<span class="kw">use</span> <span class="ident">bio::data_structures::fmindex</span>::{<span class="ident">BackwardSearchResult</span>, <span class="ident">FMIndex</span>, <span class="ident">FMIndexable</span>};
<span class="kw">use</span> <span class="ident">bio::data_structures::suffix_array::suffix_array</span>;
<span class="kw">use</span> <span class="ident">bio::io::fastq</span>;
<span class="kw">use</span> <span class="ident">bio::io::fastq::FastqRead</span>;
<span class="kw">use</span> <span class="ident">std::io</span>;

<span class="comment">// a given text</span>
<span class="kw">let</span> <span class="ident">text</span> <span class="op">=</span> <span class="string">b&quot;ACAGCTCGATCGGTA$&quot;</span>;
<span class="kw">let</span> <span class="ident">pattern</span> <span class="op">=</span> <span class="string">b&quot;ATCG&quot;</span>;

<span class="comment">// Create an FM-Index for the given text.</span>

<span class="comment">// instantiate an alphabet</span>
<span class="kw">let</span> <span class="ident">alphabet</span> <span class="op">=</span> <span class="ident">alphabets::dna::iupac_alphabet</span>();
<span class="comment">// calculate a suffix array</span>
<span class="kw">let</span> <span class="ident">sa</span> <span class="op">=</span> <span class="ident">suffix_array</span>(<span class="ident">text</span>);
<span class="comment">// calculate the Burrows-Wheeler-transform</span>
<span class="kw">let</span> <span class="ident">bwt</span> <span class="op">=</span> <span class="ident">bwt</span>(<span class="ident">text</span>, <span class="kw-2">&amp;</span><span class="ident">sa</span>);
<span class="comment">// calculate the vectors less and Occ (occurrences)</span>
<span class="kw">let</span> <span class="ident">less</span> <span class="op">=</span> <span class="ident">less</span>(<span class="kw-2">&amp;</span><span class="ident">bwt</span>, <span class="kw-2">&amp;</span><span class="ident">alphabet</span>);
<span class="kw">let</span> <span class="ident">occ</span> <span class="op">=</span> <span class="ident">Occ::new</span>(<span class="kw-2">&amp;</span><span class="ident">bwt</span>, <span class="number">3</span>, <span class="kw-2">&amp;</span><span class="ident">alphabet</span>);
<span class="comment">// set up FMIndex</span>
<span class="kw">let</span> <span class="ident">fmindex</span> <span class="op">=</span> <span class="ident">FMIndex::new</span>(<span class="kw-2">&amp;</span><span class="ident">bwt</span>, <span class="kw-2">&amp;</span><span class="ident">less</span>, <span class="kw-2">&amp;</span><span class="ident">occ</span>);
<span class="comment">// do a backwards search for the pattern</span>
<span class="kw">let</span> <span class="ident">interval</span> <span class="op">=</span> <span class="ident">fmindex</span>.<span class="ident">backward_search</span>(<span class="ident">pattern</span>.<span class="ident">iter</span>());
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">partial_match_len</span> <span class="op">=</span> <span class="number">0</span>;
<span class="comment">// get the locations where the pattern matched (completely in this case).</span>
<span class="kw">let</span> <span class="ident">positions</span> <span class="op">=</span> <span class="kw">match</span> <span class="ident">interval</span> {
    <span class="ident">BackwardSearchResult::Complete</span>(<span class="ident">saint</span>) =&gt; <span class="ident">saint</span>.<span class="ident">occ</span>(<span class="kw-2">&amp;</span><span class="ident">sa</span>),
    <span class="ident">BackwardSearchResult::Partial</span>(<span class="ident">saint</span>, <span class="ident">l</span>) =&gt; {
        <span class="ident">partial_match_len</span> <span class="op">=</span> <span class="ident">l</span>;
        <span class="ident">saint</span>.<span class="ident">occ</span>(<span class="kw-2">&amp;</span><span class="ident">sa</span>)
    }
    <span class="ident">BackwardSearchResult::Absent</span> =&gt; <span class="ident">Vec::new</span>(),
};
<span class="comment">// Iterate over a FASTQ file, use the alphabet to validate read</span>
<span class="comment">// sequences and search for exact matches in the FM-Index.</span>

<span class="comment">// create FASTQ reader</span>
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">reader</span> <span class="op">=</span> <span class="ident">fastq::Reader::new</span>(<span class="ident">io::stdin</span>());
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">record</span> <span class="op">=</span> <span class="ident">fastq::Record::new</span>();
<span class="kw">let</span> <span class="kw-2">mut</span> <span class="ident">partial_match_len</span> <span class="op">=</span> <span class="number">0</span>;
<span class="ident">reader</span>.<span class="ident">read</span>(<span class="kw-2">&amp;mut</span> <span class="ident">record</span>).<span class="ident">expect</span>(<span class="string">&quot;Failed to parse record&quot;</span>);
<span class="kw">while</span> <span class="op">!</span><span class="ident">record</span>.<span class="ident">is_empty</span>() {
    <span class="kw">let</span> <span class="ident">check</span> <span class="op">=</span> <span class="ident">record</span>.<span class="ident">check</span>();
    <span class="kw">if</span> <span class="ident">check</span>.<span class="ident">is_err</span>() {
        <span class="macro">panic!</span>(<span class="string">&quot;I got a rubbish record!&quot;</span>)
    }
    <span class="comment">// obtain sequence</span>
    <span class="kw">let</span> <span class="ident">seq</span> <span class="op">=</span> <span class="ident">record</span>.<span class="ident">seq</span>();
    <span class="comment">// check, whether seq is in the expected alphabet</span>
    <span class="kw">if</span> <span class="ident">alphabet</span>.<span class="ident">is_word</span>(<span class="ident">seq</span>) {
        <span class="kw">let</span> <span class="ident">interval</span> <span class="op">=</span> <span class="ident">fmindex</span>.<span class="ident">backward_search</span>(<span class="ident">seq</span>.<span class="ident">iter</span>());
        <span class="comment">// get the positions where seq matched completely</span>
        <span class="comment">// or where the maximal matching suffix of seq occurred.</span>
        <span class="kw">let</span> <span class="ident">positions</span> <span class="op">=</span> <span class="kw">match</span> <span class="ident">interval</span> {
            <span class="ident">BackwardSearchResult::Complete</span>(<span class="ident">saint</span>) =&gt; <span class="ident">saint</span>.<span class="ident">occ</span>(<span class="kw-2">&amp;</span><span class="ident">sa</span>),
            <span class="ident">BackwardSearchResult::Partial</span>(<span class="ident">saint</span>, <span class="ident">l</span>) =&gt; {
                <span class="ident">partial_match_len</span> <span class="op">=</span> <span class="ident">l</span>;
                <span class="ident">saint</span>.<span class="ident">occ</span>(<span class="kw-2">&amp;</span><span class="ident">sa</span>)
            }
            <span class="ident">BackwardSearchResult::Absent</span> =&gt; <span class="ident">Vec::new</span>(),
        };
    }
    <span class="ident">reader</span>.<span class="ident">read</span>(<span class="kw-2">&amp;mut</span> <span class="ident">record</span>).<span class="ident">expect</span>(<span class="string">&quot;Failed to parse record&quot;</span>);
}</code></pre></div>
<p>Documentation and further examples for each module can be found in the module descriptions below.</p>
<h3 id="example-multithreaded"><a href="#example-multithreaded">Example: Multithreaded</a></h3>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use</span> <span class="ident">bio::alphabets</span>;
<span class="kw">use</span> <span class="ident">bio::data_structures::bwt</span>::{<span class="ident">bwt</span>, <span class="ident">less</span>, <span class="ident">Occ</span>};
<span class="kw">use</span> <span class="ident">bio::data_structures::fmindex</span>::{<span class="ident">BackwardSearchResult</span>,<span class="ident">FMIndex</span>, <span class="ident">FMIndexable</span>};
<span class="kw">use</span> <span class="ident">bio::data_structures::suffix_array::suffix_array</span>;
<span class="kw">use</span> <span class="ident">std::sync::Arc</span>;
<span class="kw">use</span> <span class="ident">std::thread</span>;

<span class="kw">let</span> <span class="ident">text</span> <span class="op">=</span> <span class="string">b&quot;ACGGATGCTGGATCGGATCGCGCTAGCTA$&quot;</span>;
<span class="kw">let</span> <span class="ident">patterns</span> <span class="op">=</span> <span class="macro">vec!</span>[<span class="string">b&quot;ACCG&quot;</span>, <span class="string">b&quot;TGCT&quot;</span>];

<span class="comment">// Create an FM-Index for a given text.</span>
<span class="kw">let</span> <span class="ident">alphabet</span> <span class="op">=</span> <span class="ident">alphabets::dna::iupac_alphabet</span>();
<span class="kw">let</span> <span class="ident">sa</span> <span class="op">=</span> <span class="ident">suffix_array</span>(<span class="ident">text</span>);
<span class="kw">let</span> <span class="ident">bwt</span> <span class="op">=</span> <span class="ident">Arc::new</span>(<span class="ident">bwt</span>(<span class="ident">text</span>, <span class="kw-2">&amp;</span><span class="ident">sa</span>));
<span class="kw">let</span> <span class="ident">less</span> <span class="op">=</span> <span class="ident">Arc::new</span>(<span class="ident">less</span>(<span class="ident">bwt</span>.<span class="ident">as_ref</span>(), <span class="kw-2">&amp;</span><span class="ident">alphabet</span>));
<span class="kw">let</span> <span class="ident">occ</span> <span class="op">=</span> <span class="ident">Arc::new</span>(<span class="ident">Occ::new</span>(<span class="ident">bwt</span>.<span class="ident">as_ref</span>(), <span class="number">3</span>, <span class="kw-2">&amp;</span><span class="ident">alphabet</span>));
<span class="kw">let</span> <span class="ident">fmindex</span> <span class="op">=</span> <span class="ident">Arc::new</span>(<span class="ident">FMIndex::new</span>(<span class="ident">bwt</span>, <span class="ident">less</span>, <span class="ident">occ</span>));

<span class="comment">// Spawn threads to perform backward searches for each interval</span>
<span class="kw">let</span> <span class="ident">interval_calculators</span> <span class="op">=</span> <span class="ident">patterns</span>
    .<span class="ident">into_iter</span>()
    .<span class="ident">map</span>(<span class="op">|</span><span class="ident">pattern</span><span class="op">|</span> {
        <span class="kw">let</span> <span class="ident">fmindex</span> <span class="op">=</span> <span class="ident">fmindex</span>.<span class="ident">clone</span>();
        <span class="ident">thread::spawn</span>(<span class="kw">move</span> <span class="op">|</span><span class="op">|</span> <span class="ident">fmindex</span>.<span class="ident">backward_search</span>(<span class="ident">pattern</span>.<span class="ident">iter</span>()))
    })
    .<span class="ident">collect</span>::<span class="op">&lt;</span><span class="ident">Vec</span><span class="op">&lt;</span><span class="kw">_</span><span class="op">&gt;</span><span class="op">&gt;</span>();

<span class="comment">// Loop through the results, extracting the positions array for each pattern</span>
<span class="kw">for</span> <span class="ident">interval_calculator</span> <span class="kw">in</span> <span class="ident">interval_calculators</span> {
    <span class="kw">let</span> <span class="ident">positions</span> <span class="op">=</span> <span class="kw">match</span> <span class="ident">interval_calculator</span>.<span class="ident">join</span>().<span class="ident">unwrap</span>() {
        <span class="ident">BackwardSearchResult::Complete</span>(<span class="ident">saint</span>) =&gt; <span class="ident">saint</span>.<span class="ident">occ</span>(<span class="kw-2">&amp;</span><span class="ident">sa</span>),
        <span class="kw">_</span> =&gt; <span class="ident">Vec::new</span>()
    };
}</code></pre></div>
<p>Documentation and further examples for each module can be found in the module descriptions below.</p>
<h2 id="benchmarks"><a href="#benchmarks">Benchmarks</a></h2>
<p>Since Rust-Bio is based on a compiled language, similar performance to C/C++ based libraries can be expected. Indeed, we find the pattern matching algorithms of Rust-Bio to perform in the range of the C++ library Seqan:</p>
<div><table><thead><tr><th>Algorithm</th><th style="text-align: right">Rust-Bio</th><th style="text-align: right">Seqan</th></tr></thead><tbody>
<tr><td>BNDM</td><td style="text-align: right">77ms</td><td style="text-align: right">80ms</td></tr>
<tr><td>Horspool</td><td style="text-align: right">122ms</td><td style="text-align: right">125ms</td></tr>
<tr><td>BOM</td><td style="text-align: right">103ms</td><td style="text-align: right">107ms</td></tr>
<tr><td>Shift-And</td><td style="text-align: right">241ms</td><td style="text-align: right">545ms</td></tr>
</tbody></table>
</div>
<p>We measured 10000 iterations of searching pattern <code>GCGCGTACACACCGCCCG</code> in the sequence of the hg38 MT chromosome.
Initialization time of each algorithm for the given pattern was included in each iteration. Benchmarks were conducted with <em>Cargo bench</em> for Rust-Bio and <em>Python timeit</em> for Seqan on an Intel Core i5-3427U CPU.
Benchmarking Seqan from <em>Python timeit</em> entails an overhead of 1.46ms for calling a C++ binary. This overhead was subtracted from above Seqan run times.
Note that this benchmark only compares the two libraries to exemplify that Rust-Bio has comparable speed to C++ libraries: all used algorithms have their advantages for specific text and pattern structures and lengths (see <a href="https://docs.rs/bio/0.28.2/bio/pattern_matching/index.html">the pattern matching section in the documentation</a>)./!</p>
</div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="mod" href="alignment/index.html" title="bio::alignment mod">alignment</a></div><div class="item-right docblock-short"><p>Various alignment and distance computing algorithms.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="alphabets/index.html" title="bio::alphabets mod">alphabets</a></div><div class="item-right docblock-short"><p>Implementation of alphabets and useful utilities.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="data_structures/index.html" title="bio::data_structures mod">data_structures</a></div><div class="item-right docblock-short"><p>Various useful data structures.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="io/index.html" title="bio::io mod">io</a></div><div class="item-right docblock-short"><p>Readers and writers for common bioinformatics file formats.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="pattern_matching/index.html" title="bio::pattern_matching mod">pattern_matching</a></div><div class="item-right docblock-short"><p>This module contains various useful pattern matching algorithms.
The implementations are based on the lecture notes
“Algorithmen auf Sequenzen”, Kopczynski, Marschall, Martin and Rahmann, 2008 - 2015.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="scores/index.html" title="bio::scores mod">scores</a></div><div class="item-right docblock-short"></div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="seq_analysis/index.html" title="bio::seq_analysis mod">seq_analysis</a></div><div class="item-right docblock-short"><p>Sequence analysis algorithms.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="stats/index.html" title="bio::stats mod">stats</a></div><div class="item-right docblock-short"><p>Mathematical and statistical tools.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="mod" href="utils/index.html" title="bio::utils mod">utils</a></div><div class="item-right docblock-short"><p>Common utilities.</p>
</div></div></div></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="bio" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.62.1 (e092d0b6b 2022-07-16)" ></div>
</body></html>